libraries/pbrlib/genglsl/lib/mx_microfacet_specular.glsl

```diff
@@ -168,3 +168,219 @@ vec3 mx_fresnel_conductor(float cosTheta, vec3 n, vec3 k)
+const mat3 XYZ_TO_RGB = mat3(2.3706743, -0.5138850, 0.0052982, -0.9000405, 1.4253036, -0.0146949, -0.4706338, 0.0885814, 1.0093968);
+
+// Depolarization functions for natural light
+float mx_depolarize(vec2 v)
+{
+    return 0.5 * (v.x + v.y);
+}
+vec3 mx_depolarize(vec3 s, vec3 p)
+{
+    return 0.5 * (s + p);
+}
+
+// Evaluation XYZ sensitivity curves in Fourier space
+vec3 mx_eval_sensitivity(float opd, float shift)
+{
+    // Use Gaussian fits, given by 3 parameters: val, pos and var
+    float phase = 2*M_PI * opd;
+    vec3 val = vec3(5.4856e-13, 4.4201e-13, 5.2481e-13);
+    vec3 pos = vec3(1.6810e+06, 1.7953e+06, 2.2084e+06);
+    vec3 var = vec3(4.3278e+09, 9.3046e+09, 6.6121e+09);
+    vec3 xyz = val * sqrt(2*M_PI * var) * cos(pos * phase + shift) * exp(- var * phase*phase);
+    xyz.x   += 9.7470e-14 * sqrt(2*M_PI * 4.5282e+09) * cos(2.2399e+06 * phase + shift) * exp(- 4.5282e+09 * phase*phase);
+    return xyz / 1.0685e-7;
+}
+
+// A Practical Extension to Microfacet Theory for the Modeling of Varying Iridescence
+// https://belcour.github.io/blog/research/2017/05/01/brdf-thin-film.html
+vec3 mx_fresnel_airy(float cosTheta, vec3 ior, vec3 extinction, float tf_thickness, float tf_ior)
+{
+    // Convert nm -> m
+    float d = tf_thickness * 1.0e-9;
+
+    // Assume vacuum on the outside
+    float eta1 = 1.0;
+    float eta2 = tf_ior;
+
+    // Optical path difference
+    float cosTheta2 = sqrt(1.0 - mx_square(eta1/eta2) * (1.0 - mx_square(cosTheta)));
+    float D = 2.0 * eta2 * d * cosTheta2;
+
+    // First interface
+    vec2 R12, phi12;
+    mx_fresnel_dielectric_polarized(cosTheta, eta1, eta2, R12, phi12);
+    vec2 R21  = R12;
+    vec2 T121 = vec2(1.0) - R12;
+    vec2 phi21 = vec2(M_PI) - phi12;
+
+    // Second interface
+    vec2 R23, phi23;
+    mx_fresnel_conductor_polarized(cosTheta2, eta2, ior.x, extinction.x, R23, phi23);
+
+    // Phase shift
+    vec2 phi2 = phi21 + phi23;
+
+    // Compound terms
+    vec3 R = vec3(0.0);
+    vec2 R123 = R12*R23;
+    vec2 r123 = sqrt(R123);
+    vec2 Rs   = mx_square(T121)*R23 / (1-R123);
+
+    // Reflectance term for m=0 (DC term amplitude)
+    vec2 C0 = R12 + Rs;
+    vec3 S0 = mx_eval_sensitivity(0.0, 0.0);
+    R += mx_depolarize(C0) * S0;
+
+    // Reflectance term for m>0 (pairs of diracs)
+    vec2 Cm = Rs - T121;
+    for (int m=1; m<=3; ++m)
+    {
+        Cm *= r123;
+        vec3 SmS = 2.0 * mx_eval_sensitivity(m*D, m*phi2.x);
+        vec3 SmP = 2.0 * mx_eval_sensitivity(m*D, m*phi2.y);
+        R += mx_depolarize(Cm.x*SmS, Cm.y*SmP);
+    }
+
+    // Convert back to RGB reflectance
+    R = clamp(XYZ_TO_RGB * R, vec3(0.0), vec3(1.0));
+
+    return R;
+}
+
+// Parameters for Fresnel calculations.
+struct FresnelData
+{
+    int model;
+    vec3 ior;        // In Schlick Fresnel mode these two
+    vec3 extinction; // hold F0 and F90 reflectance values
+    float exponent;
+    float tf_thickness;
+    float tf_ior;
+};
```

Comment:
Refactor of how fresnel is setup and calculated, with better encapsulation, since we now have four different fresnel models to support in various places.